w11b01/test/pgdp/pengusurvivors/ResidualGraphTest.java

```diff
@@ -0,0 +1,178 @@
+                                b -> t;
+                                c -> t;
+                                """,
+                        "s", "t", List.of(List.of("s", "a", "b", "t"), List.of("s", "a", "c", "t"), List.of("s", "a", "b", "c", "t"))
+                ),
+                Arguments.of(
+                        """
+                                a -> a;
+                                """
+                        , "a", "a", List.of(List.of("a"))
+                ),
+                Arguments.of(
+                        """
+                                a -> b (1/1);
+                                b -> (0/2);
+                                """,
+                        "a", "c", List.of()
+                )
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource("getUpdateNetworkArguments")
+    void updateNetwork(List<String> vertexPath, int augFlow, String source, String expected, String expectedResidual) {
+        var sourceGraph = GraphGenerator.generateGraph(source).graph();
+        var expectedGraph = GraphGenerator.generateGraph(expected).graph();
+        var expectedResidualGraph = GraphGenerator.generateGraph(expectedResidual).graph();
+        sourceGraph.generateResidualGraph();
+        var path = GraphUtils.generatePath(sourceGraph, vertexPath);
+        sourceGraph.updateNetwork(path, augFlow);
+        assertTrue(GraphUtils.residualEquals(sourceGraph, expectedResidualGraph));
+        assertTrue(GraphUtils.normalEquals(sourceGraph, expectedGraph));
+    }
+
+    static Stream<Arguments> getUpdateNetworkArguments() {
+        return Stream.of(
+                Arguments.of(
+                        List.of("a", "b", "c"),
+                        2,
+                        """
+                                a -> b (4/10);
+                                b -> c (2/6);
+                                """,
+                        """
+                                a -> b (6/10);
+                                b -> c (4/6);
+                                """,
+                        """
+                                a -> b (0/4);
+                                b -> a (0/6);
+                                b -> c (0/2);
+                                c -> b (0/4);
+                                """
+                )
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource("getMaxFlowArguments")
+    void getMaxFlow(String graph, String source, String sink, int expected) {
+        var sourceGraph = GraphGenerator.generateGraph(graph, source, sink).graph();
+        assertEquals(expected, sourceGraph.computeMaxFlowValue());
+    }
+
+    static Stream<Arguments> getMaxFlowArguments() {
+        return Stream.of(
+                Arguments.of(
+                        """
+                                a -> b (4/10);
+                                b -> c (2/6);
+                                """,
+                        "a", "c",
+                        6
```

Comment:
The expected flowrate should be 8.

```suggestion
                        8
```